polynote-kernel/src/main/scala/polynote/kernel/PolyKernel.scala

```diff
@@ -29,78 +31,109 @@ import scala.tools.nsc.interactive.Global
 
 class PolyKernel private[kernel] (
   private val getNotebook: () => IO[Notebook],
-  val globalInfo: GlobalInfo,
+  val kernelContext: KernelContext,
   val outputDir: AbstractFile,
   dependencies: Map[String, List[(String, File)]],
   val statusUpdates: Publish[IO, KernelStatusUpdate],
-  subKernels: Map[String, LanguageKernel.Factory[IO]] = Map.empty
-) extends Kernel[IO] {
+  subKernels: Map[String, LanguageInterpreter.Factory[IO]] = Map.empty
+) extends KernelAPI[IO] {
 
   protected val logger: Logger = getLogger
 
   protected implicit val contextShift: ContextShift[IO] = IO.contextShift(ExecutionContext.fromExecutorService(Executors.newCachedThreadPool()))
 
-  private val launchingKernel = Semaphore[IO](1).unsafeRunSync()
-  private val kernels = new ConcurrentHashMap[String, LanguageKernel[IO]]()
+  private val launchingInterpreter = Semaphore[IO](1).unsafeRunSync()
+  private val interpreters = new ConcurrentHashMap[String, LanguageInterpreter[IO]]()
 
 
   /**
     * The symbol table, which stores all the currently existing runtime values
     */
-  protected lazy val symbolTable = new RuntimeSymbolTable(globalInfo, statusUpdates)
+  protected lazy val symbolTable = new RuntimeSymbolTable(kernelContext, statusUpdates)
 
   /**
     * The task queue, which tracks currently running and queued kernel tasks (i.e. cells to run)
     */
   protected lazy val taskQueue: TaskQueue = TaskQueue(statusUpdates).unsafeRunSync()
 
   // TODO: duplicates logic from runCell
-  private def runPredef(kernel: LanguageKernel[IO], language: String): IO[Unit] =
-    kernel.predefCode.fold(kernel.init()) {
-      code => taskQueue.runTaskIO(s"Predef $language", s"Predef ($language)")(_ => kernel.runCode("Predef", Nil, Nil, code)).flatMap {
-        results => results.evalMap {
-          case WrapSymbol(symbol) =>
-            symbolTable.publishAll(symbolTable.RuntimeValue.fromSymbolDecl(symbol).toList)
-          case _ => IO.unit
-        }.compile.drain
-      }
-    }
+  private def runPredef(interp: LanguageInterpreter[IO], language: String): IO[Stream[IO, Result]] =
+    interp.init() *> {
```

Comment:
I think this is fine for now but in the future it might be useful to either:
- decouple initializing the interpreter from running the predef 
- or, get rid of the distinction entirely.

Reply:
👍 agreed